### PR TITLE
Added Multi Trajectory context window generation

### PR DESCRIPTION
**Pull request by Danfoa**
I modified `TrajectoryContextDataset` to accept multiple trajectory data. 

As the data stream remains agnostic to the dimensions of the features, I had to add a "multi_traj" boolean argument at the class's initializer to assert if the first dimension of the vector is considered as the number of trajectories or not. 

I suggest enforcing a prefix input expected shape of (n_trajs, n_frames, *features_dims), to avoid this new unnecessary argument and rely on the auxiliary methods `traj_to_contexts` and `multi_traj_to_context` to deal with the more flexible shapes. 

Comments: There seems to be a perhaps unnecessarily deep hierarchy or classes `TrajectoryContextDataset` -> `TensorContextDataset` -> `ContextWindowDataset` -> `ContextWindow` -> `Sequence`, for handling the time series data slicing. Several of these classes repeat data parsing to appropriate backends and shapes. This deep hierarchy makes it quite cumbersome to modify the data pipeline for someone unfamiliar with the code. I propose to reduce this hierarchy to 1 or 2 classes maximum.

### CODE REVIEW

**Comment by Danfoa** [comment]
It's worth noting that the HuggingFace `Dataset` class accepts an ArrayLike structure with the shape (A (n_samples, context_len, *features)) consisting of either `np.ndarray` or `torch.Tensors`. It seamlessly converts the data format to the appropriate backend (numpy, torch, jax, TensorFlow) using the `dataset.with_format("tf/torch")` method.

Therefore, Kooplearn could primarily focus on trajectory/context slicing, leaving the backend management to the HuggingFace backend, which is already integrated within the framework.

File: kooplearn/data.py [hunk 1]
```diff
@@ -113,124 +110,159 @@ def __init__(
...
-                )
+                data_per_traj.append(data)
+                idx_map_per_traj.append(idx_map)
 
-        elif backend == "torch":
-            if torch is None:
-                raise ImportError(
-                    "You selected the 'torch' backend, but kooplearn wasn't able to import it."
-                )
+            if isinstance(data_per_traj[0], np.ndarray):
```
[the remark this answers is not in the document]
Comment (Danfoa) [reply kooplearn/data.py:133]: It was unclear for me from the documentation and docstring what the idx_map role was. 

I follow now what you mean. Will update.

File: kooplearn/data.py [hunk 2]
```diff
@@ -71,38 +67,39 @@ def __getitem__(self, idx):
...
+        multi_traj: bool = False,
         backend: str = "auto",
         **backend_kw,
     ):
         """
         Initializes the TrajectoryContextDataset. It takes as input a trajectory and returns a sequence of context windows.
 
         Args:
-            trajectory (ArrayLike): A trajectory of shape ``(n_frames, *features_shape)``.
+            trajectory (ArrayLike): A trajectory of shape ``(n_frames, *features_shape)`` or ``(n_trajs, n_frames, *features_shape)`` if `multi_traj` is set to ``True``.
```
[the remark this answers is not in the document]
Comment (Danfoa) [reply kooplearn/data.py:99]: I will update `multi_traj_to_context` to get a Sequence/list of arrays of `(any_number_frames, *features_shape)`. Then get the samples per each.

File: kooplearn/data.py [hunk 3]
```diff
@@ -60,12 +63,18 @@ def __init__(self, data: ArrayLike, backend: str = "auto", **backend_kw):
...
-    def __getitem__(self, idx):
+    def __getitem__(self, idx) -> 'TensorContextDataset':
         if np.issubdtype(type(idx), np.integer):
+            # TODO: The default collect behaviour is to return a list of [type] objects. This additional dimension here
+            # seems to be introduced only for a very specific customm collect_fn.
             return TensorContextDataset(self.data[idx][None, ...])
         elif isinstance(idx, slice):
             return TensorContextDataset(self.data[idx])
 
+    def __getitems__(self, indices: list[int]) -> 'TensorContextDataset':
```
[the remark this answers is not in the document]
Comment (Danfoa) [reply kooplearn/data.py:74]: Yes, this is torch specific. The `__getitems__` method is called on the [default data fetch method used by DataLoaders](https://github.com/pytorch/pytorch/blob/68cffd19f650e5bb4ff74016d709fed18e9e24ad/torch/utils/data/_utils/fetch.py#L45-L54)

**Comment by pietronvll** [comment]
@Danfoa any updates on this pull request?

**Pull request merged**